Have animations more closely match the HTML spec

mrobinson opened the pull request:
These two commits do two major things:

**Have animations ticks trigger a restyle**: This corrects synchronization issues with animations,
where the values used in layout are out of sync with what is returned by `getComputedStyle`.

**Tick the animation timer in script according to spec**: This greatly reduces the flakiness of
animation and transitions tests.

Fixes #13865.

<!-- Please describe your changes on the following line: -->


---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `___` with appropriate data: -->
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes fix #13865
- [x] There are tests for these changes

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

mrobinson committed:
Use a restyle for animation ticks

This change corrects synchronization issues with animations, by
reworking the animation processing model to do a quick restyle and
incremental layout when ticking animations.

While this change adds overhead to animation ticks, the idea is that
this will be the fallback when synchronous behavior is required to
fulfill specification requirements. In the optimistic case, many
animations could be updated and applied off-the-main-thread and then
resynchronized when style information is queried by script.

Fixes #13865.

highfive commented:
Heads up! This PR modifies the following files:
 * @asajeffrey: components/script/dom/window.rs, components/script/dom/testbinding.rs, components/script/dom/webidls/TestBinding.webidl, components/script/script_thread.rs, components/script/dom/node.rs and 1 more
 * @cbrewster: components/constellation/constellation.rs
 * @kichjang: components/script/dom/window.rs, components/script_traits/lib.rs, components/script/dom/testbinding.rs, components/script/dom/webidls/TestBinding.webidl, components/script/script_thread.rs and 2 more
 * @emilio: components/layout/animation.rs, components/style/matching.rs, components/style/animation.rs, components/style/context.rs

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify **unsafe code**. Please review it carefully!

mrobinson commented:
@bors-servo try=wpt

bors-servo commented:
:hourglass: Trying commit 04d8e279d617c5b71ffe3a0568365a09ffdfc120 with merge 101c6e724644154d6c3681db2aceb263eb8a6c4d...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/TZ6HwINmRfGtjtbi9XVtJg)

jdm commented:
Should this PR enable the disabled tests, or are there other prerequisites?

mrobinson commented:
@jdm, yep. Currently, I think there are only two disabled:
 * mozilla/css/animations/basic-transition.html
 * mozilla/css/animations/transition-raf.html

These are enabled by [removing the ini files](https://github.com/servo/servo/pull/26407/files#diff-3101d239acbaeb757c93c981b546d585) which were disabling them. Could it be that there are more disabled in another directory?

mrobinson commented:
@bors-servo try=wpt

Removing two expectations for newly passing tests.

bors-servo commented:
:hourglass: Trying commit 8131ee366a488c5b0ac7de2cf64597bc666459e0 with merge 5875d050b72204cd5c0875accce3ceacf354ce49...

jdm commented:
What about the animation-related test directories skipped in include.ini?

bors-servo commented:
:sunny: Test successful - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/OBqfexmSQemAxedsgnDEVg)
State: approved= try=True

mrobinson commented:
I wasn't sure if it was better to unskip intermittent tests later, but I'm happy to give it a shot in this change.

mrobinson commented:
It looks like there is more required to eliminate the flakiness from animations and transitions tests. My suspicion is that this is #19242 and perhaps, in addition, making sure that transition and animation events arrive to the script thread before the next animation tick is triggered. I'll start taking a look at this tomorrow. It seems like it would be good to have this all sorted out before going too much further.

mrobinson commented on components/script/dom/node.rs:
```diff
@@ -170,13 +170,13 @@ bitflags! {
-        const CLICK_IN_PROGRESS = 1 << 2;
+        const CLICK_IN_PROGRESS = 1 << 3;
 
         #[doc = "Specifies whether this node is focusable and whether it is supposed \
                  to be reachable with using sequential focus navigation."]
-        const SEQUENTIALLY_FOCUSABLE = 1 << 3;
+        const SEQUENTIALLY_FOCUSABLE = 1 << 4;
 
-        // There are two free bits here.
+        // There is one bit free here.
```
Whoops. This is left over from a previous version of this change where I added `NodeFlags::HAS_ANIMATION_ONLY_DIRTY_DESCENDANTS`. It turns out that isn't necessary for this change, so I'll remove these changes as well.

mrobinson committed:
Start having animations conform to the HTML spec

This is a small step toward fixing #19242. The main idea is that the
clock for animations should advance as the event loop ticks. We
accomplish this by moving the clock from layout and naming it the
"animation timeline" which is the spec language. This should fix
flakiness with animations and transitions tests where a reflow could
move animations forward while script was running.

This change also starts to break out transition and animation events
into their own data structure, because it's quite likely that the next
step in fixing #19242 is to no longer send these events through a
channel.

mrobinson commented:
@bors-servo try=wpt

bors-servo commented:
:hourglass: Trying commit 7ff890041ae8c35e2ee6749f1093493ea062b466 with merge bf502419bca064c5eff93e3d10b63e135a78adfd...

bors-servo commented:
:hourglass: Trying commit 3f3171e0c78b7d2f4a0d63e15a773141a20999e4 with merge 78a44128094b23b7d5840bcaf84b16ea899d0db3...

jdm commented:
I'm liking these changes more and more!

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/J9yA-MJiTkqKwrftuXjpXA)

mrobinson commented:
@bors-servo try=wpt

bors-servo commented:
:hourglass: Trying commit 3a74013abcec241d67d2685e52a031409dc59dd4 with merge 47ba832de2b3d2e553ddd1f1db80729ae3fd3866...

bors-servo commented:
:sunny: Test successful - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/LU9kiz66R_Sd9lspKobArg)
State: approved= try=True

mrobinson commented:
@bors-servo try=wpt-mac

jdm commented:
@bors-servo retry

bors-servo commented:
:hourglass: Trying commit 3a74013abcec241d67d2685e52a031409dc59dd4 with merge 473e36821df7f2bd8afd3e2dc13554cce89a6022...

bors-servo commented:
:sunny: Test successful - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/bw5oDwkgRi26c4bdSXHsag)
State: approved= try=True

mrobinson commented:
@bors-servo r=jdm

bors-servo commented:
:pushpin: Commit 3a74013 has been approved by `jdm`

<!-- @bors-servo r=jdm 3a74013abcec241d67d2685e52a031409dc59dd4 -->

bors-servo commented:
:hourglass: Testing commit 3a74013abcec241d67d2685e52a031409dc59dd4 with merge b290ad95c159aeefa6e3fca91876e2a5fc584584...

bors-servo commented:
:sunny: Test successful - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/Vw6NfDXmQCmU1UW7Yydq6w)
Approved by: jdm
Pushing b290ad95c159aeefa6e3fca91876e2a5fc584584 to master...
